### PR TITLE
allow more args in run task

### LINKED ISSUE #153145: Add command for filtering task quick open entries
Currently when you run command `workbench.action.tasks.runTask` without providing a task name it displays the task quick open menu. Would it be possible to add a command that displays the task quick open with a filtered list of tasks eg only show tasks from one task provider.

### LINKED ISSUE #154201: Allow an object for arguments in runTask
We can use this keybinding currently:

```json
{ "key": "ctrl+space", "command": "workbench.action.tasks.runTask", "args": "Core - Build" }
```

But args doesn't give any hints to the user on how to use it, we should change to _also_ support this and add a schema:

```json
{ "key": "ctrl+space", "command": "workbench.action.tasks.runTask", "args": { "taskName": "Core - Build" } }
``

Related: https://github.com/microsoft/vscode/issues/153145

### PR DESCRIPTION
**Pull request by meganrogge**
This PR fixes #154201 
fixes #153145

to do:
get intellisense to work

### CODE REVIEW

**Review by aeschli — DISMISSED** [review]

File: src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts [hunk 1]
```diff
@@ -312,9 +321,16 @@ export class TaskQuickPick extends Disposable {
...
+		picker.show();
 		picker.busy = false;
 		const secondLevelPickerResult = await new Promise<ITaskTwoLevelQuickPickEntry | undefined | null>(resolve => {
+			if (!picker) {
+				return;
+			}
 			Event.once(picker.onDidAccept)(async () => {
+				if (!picker) {
+					return;
+				}
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts:333]: way around this besides `!` ?

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 2]
```diff
@@ -2732,28 +2770,56 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
...
+				if (filter && typeof typeFilter) {
+					const taskQuickPick = new TaskQuickPick(this, this._configurationService, this._quickInputService, this._notificationService, this._themeService, this._dialogService);
+					const result = await taskQuickPick.doPickerSecondLevel(filter);
+					if (result?.task) {
+						pickThen(result.task as Task);
+						taskQuickPick.dispose();
+						return;
+					} else {
+						return;
+					}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2788]:
```suggestion
					if (result?.task) {
						pickThen(result.task as Task);
						taskQuickPick.dispose();
					}
					return;
```

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 3]
```diff
@@ -2732,28 +2770,56 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
...
 
-		this._showIgnoredFoldersMessage().then(() => {
+		this._showIgnoredFoldersMessage().then(async () => {
 			if (this._configurationService.getValue(USE_SLOW_PICKER)) {
 				let taskResult: { tasks: Promise<Task[]>; grouped: Promise<TaskMap> } | undefined = undefined;
 				if (!tasks) {
 					taskResult = this._tasksAndGroupedTasks();
 				}
+				if (filter && typeof typeFilter) {
+					const taskQuickPick = new TaskQuickPick(this, this._configurationService, this._quickInputService, this._notificationService, this._themeService, this._dialogService);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2780]: Instantiation service createInstance?

**Comment by Tyriar** [comment]
@meganrogge was the issue just that the @ decorators are missing here?

https://github.com/microsoft/vscode/blob/1baeb71f07630dbdb6745f891b03e6a86aa93fff/src/vs/workbench/contrib/tasks/browser/taskQuickPick.ts#L48-L57

**Review by Tyriar — APPROVED** [review]

**Pull request merged**